### DISCUSSION TITLE
Installazione di Caddy 2

### DISCUSSION THREAD

**Discussion by paolomorabito** [indomus/forum#841] [HUB: Home Assistant]
Ho seguito la guida su indomus per installare e configurare Caddy 2 (https://indomus.it/guide/installare-e-configurare-caddy-su-home-assistant-os-hassos-hassio/). Purtroppo però ho un problema. Il mio fritzbox non permette di fare il forward della porta 443. Inoltre non posso utilizzare la porta 80. Per questo motivo il challenge non viene effettuato correttamente. C'è un modo per utilizzare altre porte (di cui poi fare il forward)? Oppure come poter utilizzare il DNS Challenge tramite provider come DNS MADE EASY

**Comment by snakuzzo** [comment]
La challenge http-01 necessita per forza della porta 80
La challenge tls-alpn-01 necessita per forza della porta 443
Ma comunque io ho il Fritzbox ed il forward della 443 te lo fa fare. Guarda...

![01](https://user-images.githubusercontent.com/7205371/136619749-d900485f-9ad6-44bf-a8b3-bf9fdf60ed19.png)

![02](https://user-images.githubusercontent.com/7205371/136619756-cfaf4b35-444f-49a0-ac55-0537a030d69a.png)

**Comment by corri57** [comment]
Salve, e' un po di giorni che cerco di far funzionare caddy 2 ma non ne vengo a capo. Anche io ho seguito la guida all'installazione di caddy2 https://indomus.it/guide/installare-e-configurare-caddy-su-home-assistant-os-hassos-hassio/  ma non riesco, credo,  ad ottenere alcun certificato.
Il mio logger riporta:

```
s6-rc: info: service s6rc-oneshot-runner: starting
s6-rc: info: service s6rc-oneshot-runner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service legacy-cont-init: starting
cont-init: info: running /etc/cont-init.d/00-banner.sh
-----------------------------------------------------------
 Add-on: Caddy 2
 Open source web and proxy server with automatic HTTPS
-----------------------------------------------------------
 Add-on version: 1.2.5
 You are running the latest version of this add-on.
 System: Home Assistant OS 9.3  (armv7 / raspberrypi2)
 Home Assistant Core: 2022.10.5
 Home Assistant Supervisor: 2022.10.2
-----------------------------------------------------------
 Please, share the above information when looking for help
 or support in, e.g., GitHub, forums or the Discord chat.
-----------------------------------------------------------
cont-init: info: /etc/cont-init.d/00-banner.sh exited 0
cont-init: info: running /etc/cont-init.d/01-log-level.sh
Log level is set to INFO
cont-init: info: /etc/cont-init.d/01-log-level.sh exited 0
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service legacy-services: starting
services-up: info: copying legacy longrun caddy (no readiness notification)
s6-rc: info: service legacy-services successfully started
INFO: Prepare Caddy...
INFO: Use built-in Caddy
v2.6.2 h1:wKoFIxpmOJLGl3QXoo6PNbYvGW4xLEgo32GPBEjWL8o=
INFO: Prepare Caddyfile...
INFO: Caddyfile found at /config/caddy/Caddyfile
INFO: Run Caddy...
{"level":"info","ts":1667982931.7478166,"msg":"using provided configuration","config_file":"/config/caddy/Caddyfile","config_adapter":""}
{"level":"warn","ts":1667982931.7631478,"msg":"Caddyfile input is not formatted; run the 'caddy fmt' command to fix inconsistencies","adapter":"caddyfile","file":"/config/caddy/Caddyfile","line":3}
{"level":"info","ts":1667982931.7772903,"logger":"admin","msg":"admin endpoint started","address":"localhost:2019","enforce_origin":false,"origins":["//localhost:2019","//[::1]:2019","//127.0.0.1:2019"]}
{"level":"info","ts":1667982931.7800827,"logger":"http","msg":"server is listening only on the HTTPS port but has no TLS connection policies; adding one to enable TLS","server_name":"srv0","https_port":443}
{"level":"info","ts":1667982931.7805774,"logger":"http","msg":"enabling automatic HTTP->HTTPS redirects","server_name":"srv0"}
{"level":"info","ts":1667982931.7802534,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0x46360a0"}
{"level":"info","ts":1667982931.8045676,"logger":"http","msg":"enabling HTTP/3 listener","addr":":443"}
{"level":"info","ts":1667982931.8091679,"logger":"tls","msg":"cleaning storage unit","description":"FileStorage:/ssl/caddy"}
{"level":"info","ts":1667982931.8103766,"logger":"http.log","msg":"server running","name":"srv0","protocols":["h1","h2","h3"]}
{"level":"info","ts":1667982931.833417,"logger":"http.log","msg":"server running","name":"remaining_auto_https_redirects","protocols":["h1","h2","h3"]}
{"level":"info","ts":1667982931.8336554,"logger":"http","msg":"enabling automatic TLS certificate management","domains":["cbottino.duckdns.org"]}
{"level":"info","ts":1667982931.8633199,"logger":"tls","msg":"finished cleaning storage units"}
{"level":"info","ts":1667982931.9285314,"msg":"autosaved config (load with --resume flag)","file":"/data/caddy/autosave.json"}
{"level":"info","ts":1667982931.9286823,"msg":"serving initial configuration"}
{"level":"info","ts":1667982931.9301848,"logger":"watcher","msg":"watching config file for changes","config_file":"/config/caddy/Caddyfile"}
```
Per quanto riguarda le configurazioni del caddy e del Caddyfile ho:

```
args:
  - "--watch"
env_vars: []
log_level: info
non_caddyfile_config:
  destination: 192.168.1.152
  domain: casamia.duckdns.org
  email: miamail@alice.it
  port: 8123
config_path: /config/caddy/Caddyfile
```

mentre il Caddyfile:

```
#  When your config is ok comment block below to disable use of AMCE staging server
{
    acme_ca https://acme-staging-v02.api.letsencrypt.org/directory 
}

https://casamia.duckdns.org { 
  respond "Saluti dal Crew inDomus.it"
  file_server
}

```

Per il port forwarding, il fastgate non mi permette di mappare la porta 80, quindi ho mappato solo la 443 ridirigendola su 192.168.1.152 dove e' presente HA e caddy.

Sapete indirizzarmi su una possibile soluzione e/o prove da fare?
Grazie

**Reply by snakuzzo** [reply]
Se usi FW devi assicurarti di avere IP pubblico

**Reply by snakuzzo** [reply]
Inoltre, visto che usi l'addon su HA OS, ti basta questo ... senza usare alcun Caddyfile e riavviare l'addon

![image](https://user-images.githubusercontent.com/7205371/200815150-bdf4d596-be83-4ced-8472-a2295aa44f64.png)

**Comment by corri57** [comment]
Grazie del suggerimento.
Inizialmente il registro sembrava OK ma non l'avevo salvato. Poi ho ripetuto il tutto ma non credo che vada bene.
Ho fatto la configurazione suggerita, quindi no Caddyfile. Provando a collegarmi con casamia.duckdns.org ottengo
400: Bad Request (tramite chrome). il registro invece e':

```
s6-rc: info: service s6rc-oneshot-runner: starting
s6-rc: info: service s6rc-oneshot-runner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service legacy-cont-init: starting
cont-init: info: running /etc/cont-init.d/00-banner.sh
-----------------------------------------------------------
 Add-on: Caddy 2
 Open source web and proxy server with automatic HTTPS
-----------------------------------------------------------
 Add-on version: 1.2.5
 You are running the latest version of this add-on.
 System: Home Assistant OS 9.3  (armv7 / raspberrypi2)
 Home Assistant Core: 2022.10.5
 Home Assistant Supervisor: 2022.10.2
-----------------------------------------------------------
 Please, share the above information when looking for help
 or support in, e.g., GitHub, forums or the Discord chat.
-----------------------------------------------------------
cont-init: info: /etc/cont-init.d/00-banner.sh exited 0
cont-init: info: running /etc/cont-init.d/01-log-level.sh
Log level is set to INFO
cont-init: info: /etc/cont-init.d/01-log-level.sh exited 0
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service legacy-services: starting
services-up: info: copying legacy longrun caddy (no readiness notification)
s6-rc: info: service legacy-services successfully started
INFO: Prepare Caddy...
INFO: Use built-in Caddy
v2.6.2 h1:wKoFIxpmOJLGl3QXoo6PNbYvGW4xLEgo32GPBEjWL8o=
INFO: Prepare Caddyfile...
INFO: No Caddyfile found
INFO: Use non_caddyfile_config
INFO: Run Caddy...
{"level":"info","ts":1668010874.2718441,"msg":"using provided configuration","config_file":"/etc/caddy/Caddyfile","config_adapter":""}
{"level":"warn","ts":1668010874.285734,"msg":"Caddyfile input is not formatted; run the 'caddy fmt' command to fix inconsistencies","adapter":"caddyfile","file":"/etc/caddy/Caddyfile","line":2}
{"level":"info","ts":1668010874.294946,"logger":"admin","msg":"admin endpoint started","address":"localhost:2019","enforce_origin":false,"origins":["//localhost:2019","//[::1]:2019","//127.0.0.1:2019"]}
{"level":"info","ts":1668010874.3014057,"logger":"http","msg":"server is listening only on the HTTPS port but has no TLS connection policies; adding one to enable TLS","server_name":"srv0","https_port":443}
{"level":"info","ts":1668010874.3017085,"logger":"http","msg":"enabling automatic HTTP->HTTPS redirects","server_name":"srv0"}
{"level":"info","ts":1668010874.3038533,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0x2d223c0"}
{"level":"info","ts":1668010874.3081985,"logger":"http","msg":"enabling HTTP/3 listener","addr":":443"}
{"level":"info","ts":1668010874.3109396,"logger":"http.log","msg":"server running","name":"srv0","protocols":["h1","h2","h3"]}
{"level":"info","ts":1668010874.3119671,"logger":"http.log","msg":"server running","name":"remaining_auto_https_redirects","protocols":["h1","h2","h3"]}
{"level":"info","ts":1668010874.312116,"logger":"http","msg":"enabling automatic TLS certificate management","domains":["cbottino.duckdns.org"]}
{"level":"info","ts":1668010874.3142788,"logger":"tls","msg":"cleaning storage unit","description":"FileStorage:/ssl/caddy"}
{"level":"info","ts":1668010874.3380673,"logger":"tls","msg":"finished cleaning storage units"}
{"level":"info","ts":1668010874.4075594,"msg":"autosaved config (load with --resume flag)","file":"/data/caddy/autosave.json"}
{"level":"info","ts":1668010874.4077108,"msg":"serving initial configuration"}
```

Il mio FW mi da un IP statico, comunque utilizzo duckdns. 
Un dubbio, perche' posso non utilizzare Caddyfile?
Come faccio poi  se volessi gestire altri servizi anche su altri dispositivi?

**Reply by snakuzzo** [reply]
1. ip statico non vuol dire ip pubblico. Ad ogni modo, da quello che hai scritto dopo, si capisce che trattasi di ip pubblico.
2. Il 400 è dato dal fatto che devi mettere nel trusted proxy del blocco http del configuration.yaml di HA l'indirizzo del container di caddy (o comunque del proxy da cui arriva la richiesta). Se ti logghi in locale su HA e vai nei log lo trovi scritto ad ogni tentativo d'accesso che fai da fuori. Aggiungi questo dettaglio e vedrai che il problema sparirà e ti collegherai senza problemi (e comunque questa cosa l'abbiamo scritta)
3. Se un giorno dovessi aver bisogno del Caddyfile...lo crei e lo usi... (anche questo è scritto nella documentazione dell'addon)

**Comment by corri57** [comment]
Purtroppo ancora non va.
In configuration.yaml ho:

```
http:

  use_x_forwarded_for: true

  trusted_proxies: 
    - 127.0.0.1
```

Dopo una prova di collegamento dall'esterno ho ancora errore 400 e  nei registri di sistema e trovo:

```
Dettagli registro (ERRORE)
Logger: homeassistant.components.http.forwarded
Source: components/http/forwarded.py:123
Integration: HTTP ([documentation](https://www.home-assistant.io/integrations/http), [issues](https://github.com/home-assistant/home-assistant/issues?q=is%3Aissue+is%3Aopen+label%3A%22integration%3A+http%22))
First occurred: 00:07:58 (3 occurrences)
Last logged: 00:08:04

Received X-Forwarded-For header from an untrusted proxy ::1
```

**Reply by Scialla** [reply]
127.0.0.1 è l'host non l'IP del container con caddy, che trovi nel log degli errori come ha detto snakuzzo (presumo inizi con 172.x.x.x)

**Reply by snakuzzo** [reply]
No @Scialla... lì gli logga l'IP v6 della loopback.
Quindi @corri57 fai così...

```yaml
http:
  use_x_forwarded_for: true
  trusted_proxies: 
    - 127.0.0.1
    - ::1
```

**Reply by corri57** [reply]
@snakuzzo esatto, cosi funziona, grazie.
Sto cercando in rete, cosa significa "::1"?

**Reply by snakuzzo** [reply]
E' l'equivalente di 127.0.0.1 ... ma è relativo ad ipv6

**Reply by Scialla** [reply]
si scusa ho scritto host ma è chiaramente il loopback, sorry. Non avevo però pensato all'IPv6
@corri57 ::1 è l'abbreviazione di 0:0:0:0:0:0:0:1 ovvero il loopback dell'IPv6
Edit, abbiamo scritto insieme Die

**Reply by corri57** [reply]
@snakuzzo  @Scialla ok, grazie.

**Comment by corri57** [comment]
Eccomi qui ancora con problemi di configurazione caddy.
Da fuori casa riesco a raggiungere il mio HA non facendo uso di caddyfile.
Adesso vorrei vorrei far fare del port-forwarding e configurare di conseguenza il caddyfile.
Sto seguendo la stessa guida dell'installazione caddy, in particolare l'ultima parte con l'esempio di utilizzo della porta 9000 ....
La configurazione caddy e':

``` 
args: []
env_vars: []
log_level: info
non_caddyfile_config:
  destination: localhost
  domain: casamia.duckdns.org:443
  email: miaemail@alice.it
  port: 8123
config_path: /config/caddy/Caddyfile 
```

Il Caddyfile:
```
(https_header) {
  header {
    Strict-Transport-Security "max-age=31536000; includeSubdomains"
    X-XSS-Protection "1; mode=block"
    X-Content-Type-Options "nosniff"
    X-Frame-Options "SAMEORIGIN"
    Referrer-Policy "same-origin"
  }
}

#https://casamia.duckdns.org { 
#  respond "Saluti dal Crew inDomus"
#  file_server
#}
https://casamia.duckdns.org:9000 { 
  import https_header
  reverse_proxy http://192.168.1.152:8123
}
```
Su router fastgate ho aperto la porta 9000 (esterna e interna) verso 192.168.1.152.

Quando provo a collegarmi a HA ho il msg:  
```
Client sent an HTTP request to an HTTPS server.
(La tua connessione a questo sito non e' sicura ....)
```
Manca ancora qualche altra conf? o c'e' una conf errata?

**Reply by snakuzzo** [reply]
Perché ora vuoi usare il Caddyfile? Non ti bastava NON usarlo?
L'addon mette a disposizione una configurazione semplice proprio per fare quello che ti serve.

Ad ogni modo, se vuoi usarlo devi configurare cosi...

Port forwarding 9000 verso 9000 di caddy
Poi è Caddy a redirigetr sulla 8123

E la chiamata devi farla in https...tu hai fatto una chiamata in http

**Comment by corri57** [comment]
Se ho capito bene, caddy mi permetterebbe di avere connessioni sicure anche per altre cose locali fuori da HA, quindi volevo provare anche questa possibilita (che forse potrebbe anche non avere senso).

Avevo gia provato a fare la chiamata in https ma con lo stesso risultato.
Ho l'impressione che il problema sia nella configurazione delle porte su router e quindi su quali utilizzare nelle configurazioni HA.
Attualmente ho:
```
IP destinazione: 192.168.1.152
Porta esterna: TCP porta 9000
Porta interna: TCP porta 9000

IP destinazione: 192.168.1.152
Porta esterna: TCP porta 8123
Porta interna: TCP porta 8123
```
il collegamento a casamia.duckdns.org:8123 e' ok, mentre casamia.duckdns.org:9000 no.

**Reply by snakuzzo** [reply]
Perché hai Core configurato in https con i certificati probabilmente.
Devi eliminare quella parte e lasciarlo in http, specificando solo un paio di cose. Trovi tutto nella guida di Caddy

**Reply by corri57** [reply]
@snakuzzo ,purtroppo sono alle prime armi con HA,  capisco che bisogna sbatterci la testa soprattutto se le tematiche sono gia discusse/documentate, ma non riesco a seguire facilmente i tuoi suggerimenti per risolvere il problema. Potresti darmi qualche altro dettaglio delle cose da fare. Purtroppo non ricordo di questa eventuale configurazione in http/https. Configurazione che si fa dal mio hassio? oppure  da Caddy? Se va fatta da caddy, quale parte della guida dovrei vedere?

**Reply by snakuzzo** [reply]
Nel `configuration.yaml` di HA
